src/main/java/org/primefaces/component/datatable/DataTable.java

```diff
@@ -1547,50 +1411,134 @@ public void resetMultiViewState() {
     }
 
     public String getGroupedColumnIndexes() {
-        List<UIColumn> columns = getColumns();
-        int size = columns.size();
-        boolean hasIndex = false;
-        if (size > 0) {
-            StringBuilder sb = new StringBuilder();
-            sb.append("[");
-            for (int i = 0; i < size; i++) {
-                UIColumn column = columns.get(i);
-                if (column.isGroupRow()) {
-                    if (hasIndex) {
-                        sb.append(",");
-                    }
+        return IntStream.range(0, columns.size())
+                .filter(i -> columns.get(i).isGroupRow())
+                .mapToObj(Objects::toString)
+                .collect(Collectors.joining(",", "[", "]"));
+    }
 
-                    sb.append(i);
-                    hasIndex = true;
-                }
+    public String getSortMetaAsString() {
+        return getSortByAsMap().keySet().stream()
+                .collect(Collectors.joining("','", "['", "']"));
+    }
+
+    public static ValueExpression createValueExprFromVarField(FacesContext context, String var, String field) {
+        if (LangUtils.isValueBlank(var) || LangUtils.isValueBlank(field)) {
+            throw new FacesException("var and field must be non null");
+        }
+
+        return context.getApplication().getExpressionFactory()
+                .createValueExpression(context.getELContext(), "#{" + var + "." + field + "}", String.class);
+    }
+
+    protected Map<String, SortMeta> initSortBy(Object sortByTmp) {
+        Map<String, SortMeta> sortMeta = new HashMap<>();
+        boolean sorted = false;
+
+        HeaderRow headerRow = getHeaderRow();
+        if (headerRow != null) {
+            SortMeta s = SortMeta.of(getFacesContext(), getVar(), headerRow);
+            sortMeta.put(s.getColumnKey(), s);
+            sorted = true;
+        }
+
+        for (UIColumn column : getColumns()) {
+            if (column instanceof DynamicColumn) {
+                ((DynamicColumn) column).applyStatelessModel();
+            }
+
+            if (!column.isSortable()) {
+                continue;
+            }
+
+            SortMeta s = SortMeta.of(getFacesContext(), getVar(), column);
+            if (s == null) {
+                continue;
             }
-            sb.append("]");
 
-            return sb.toString();
+            sortMeta.put(s.getColumnKey(), s);
+            sorted |= s.isActive();
         }
-        return null;
+
+        setDefaultSort(sorted);
+
+        // merge internal sortBy with user sortBy
+        if (sortByTmp != null) {
+            Collection<SortMeta> sortBy;
+            if (sortByTmp instanceof SortMeta) {
+                sortBy = Collections.singletonList((SortMeta) sortByTmp);
+            }
+            else if (!(sortByTmp instanceof Collection)) {
+                throw new FacesException("DataTable#sortBy expects a single or a collection of SortMeta");
+            }
```

Comment:
Would it be theoretically possible to add there some kind of backward-compatibility? (Convert PF<=8  sortBy into SortMeta and write warning to log.)
I don´t want to say this is a good idea. Just thinking about it.

Reply:
True, on one hand it can easily be done, on the other hand, since clients will have to do a bit of cleanup of their tables it might be the best opportunity for them to change their sortBy than wait till the backward compatibility is removed ... The decision is up to @tandraschko